# Classification Model Selection by Accuracy Score
* What is the best Classfication Model for the given dataset?

> dataset - breast cancer

> class 2: N, class 4 : P

## Importing the libraries

In [9]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Importing the dataset

In [10]:
dataset = pd.read_csv('Data.csv')
x = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values
dataset.head()

Sample code number  Clump Thickness  Uniformity of Cell Size  \
0             1000025                5                        1   
1             1002945                5                        4   
2             1015425                3                        1   
3             1016277                6                        8   
4             1017023                4                        1   

   Uniformity of Cell Shape  Marginal Adhesion  Single Epithelial Cell Size  \
0                         1                  1                            2   
1                         4                  5                            7   
2                         1                  1                            2   
3                         8                  1                            3   
4                         1                  3                            2   

   Bare Nuclei  Bland Chromatin  Normal Nucleoli  Mitoses  Class  
0            1                3                1        1      2  
1           10                3                2        1      2  
2            2                3                1        1      2  
3            4                3                7        1      2  
4            1                3                1        1      2

## Splitting the dataset into the Training set and Test set

In [11]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size =0.25, random_state=0)

## Feature Scaling

In [12]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_train = sc.fit_transform(x_train)
x_test = sc.transform(x_test)

# 1. Logistic Regression

## Training the Logistic Regression model on the Training set

In [28]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(x_train,y_train)

LogisticRegression(random_state=0)

In [29]:
y_pred = classifier.predict(x_test)

## Making the Confusion Matrix
> Confusion Matrix : To decide which method should be performed with the given data sets, we need to summurize how each method performed on the testing Data. 

> Accuracy score - calculated by dividing the number of correct predictions (true positives and true negatives) by the total number of predictions.

In [30]:
# Confusion Matrix
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(y_test,y_pred)
print(matrix)

[[103   4]
 [  5  59]]


In [31]:
# Accuracy Score
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred)

0.9473684210526315

# 2. K-NN

## Training the K-NN model on the Training set

In [32]:
from sklearn.neighbors import KNeighborsClassifier
classifier2 = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
classifier2.fit(x_train, y_train)

y_pred2 = classifier2.predict(x_test)

## Finding Confusion Matrix & Accuracy Score

In [33]:
# Confusion Matrix
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(y_test,y_pred2)
print(matrix)

# Accuracy Score
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred2)

[[103   4]
 [  5  59]]


0.9473684210526315

# 3. SVM

## Training the SVM model on the Training set

In [35]:
from sklearn.svm import SVC
classifier3 = SVC(kernel = 'linear', random_state = 0)

classifier3.fit(x_train, y_train)

y_pred3 = classifier3.predict(x_test)

## Finding Confusion Matrix & Accuracy Score

In [36]:
# Confusion Matrix
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(y_test,y_pred3)
print(matrix)

# Accuracy Score
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred3)

[[102   5]
 [  5  59]]


0.9415204678362573

# 4. Kernel SVM

## Training the Kernel SVM model on the Training set

In [37]:
from sklearn.svm import SVC
classifier4 = SVC(kernel = 'rbf', random_state = 0)
classifier4.fit(x_train, y_train)

y_pred4 = classifier4.predict(x_test)

## Finding Confusion Matrix & Accuracy Score

In [38]:
# Confusion Matrix
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(y_test,y_pred4)
print(matrix)

# Accuracy Score
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred4)

[[102   5]
 [  3  61]]


0.9532163742690059

# 5. Naive Bayes

## Training the Naive Bayes model on the Training set

In [39]:
from sklearn.naive_bayes import GaussianNB
classifier5 = GaussianNB()
classifier5.fit(x_train, y_train)

y_pred5 = classifier5.predict(x_test)

## Finding Confusion Matrix & Accuracy Score

In [40]:
# Confusion Matrix
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(y_test,y_pred5)
print(matrix)

# Accuracy Score
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred5)

[[99  8]
 [ 2 62]]


0.9415204678362573

# 6. Decision Tree Classification

## Training the Decision Tree Classification model on the Training set

In [41]:
from sklearn.tree import DecisionTreeClassifier
classifier6 = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier6.fit(x_train, y_train)

y_pred6 = classifier6.predict(x_test)

## Finding Confusion Matrix & Accuracy Score

In [42]:
# Confusion Matrix
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(y_test,y_pred6)
print(matrix)

# Accuracy Score
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred6)

[[103   4]
 [  3  61]]


0.9590643274853801

# 7. Random Forest Classification

## Training the Random Forest Classification model on the Training set

In [47]:
from sklearn.ensemble import RandomForestClassifier
classifier7 = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
classifier7.fit(x_train, y_train)

y_pred7 = classifier7.predict(x_test)

## Finding Confusion Matrix & Accuracy Score

In [48]:
# Confusion Matrix
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(y_test,y_pred7)
print(matrix)

# Accuracy Score
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred7)

[[102   5]
 [  6  58]]


0.935672514619883

## Compare the Accuracy Score

1. Logistic Regression
> 0.9473684210526315

2. K-NN
> 0.9473684210526315

3. SVM
> 0.9415204678362573

4. Kernel. SVM
> 0.9532163742690059

5. Naive Bayes
> 0.9415204678362573

6. Decision Tree Classification
> 0.9590643274853801

7. Random Forest Classification
> 0.935672514619883

### Conclusion
- the best model : Decision Tree Classification model